### DISCUSSION TITLE
Could you advise on the best approach for implementing prompt=none?

### DISCUSSION THREAD

**Discussion by nannany** [zitadel/oidc#669] [Q&A] [answered]
I’m using zitadel/oidc and would like to implement support for prompt=none. When considering the implementation on the OP side for the authentication request, I expect the following behavior: if a session already exists, redirect directly to the relying party’s callback endpoint; otherwise, display the login screen.

I believe this conditional branching is best implemented in [op.Authorize](https://github.com/zitadel/oidc/blob/24869d281155064b0a4a6339ad641ba76bac6177/pkg/op/server.go#L67), which likely involves using [op.AuthResponse](https://github.com/zitadel/oidc/blob/24869d281155064b0a4a6339ad641ba76bac6177/pkg/op/auth_request.go#L454). However, op.Authorize does not seem to allow passing the necessary arguments to op.AuthResponse.

What would be the appropriate approach to handle this?

**Comment by hifabienne** [comment]
Hi @nannany 
According to the OIDC spec, the prompt=none should never display any authentication or consent screens.
If a user is not authenticated an error should be shown.
https://openid.net/specs/openid-connect-core-1_0.html
<img width="1404" alt="image" src="https://github.com/user-attachments/assets/d108075c-9d27-437a-83c9-478ccbcce52a">

**Reply by nannany** [reply]
Thank you for your response.

I understand that, according to the OIDC spec, prompt=none should never display any authentication or consent screens.

I'm interested in how to implement prompt=none using zitadel/oidc in cases where the user is already authenticated (for example, when a session is present and identifiable).

**Reply by hifabienne** [reply]
In the relying party is a function called WithPrompt, you should be able to set it with that

**Reply by nannany** [reply]
I would like to implement an OpenID Provider (OP), so this is not about the relying party.

For the prompt=none flow, I envision the process as follows:

```mermaid
sequenceDiagram
    participant u as User Agent
    participant r as Relying Party
    participant o as OpenID Provider
    u ->> r: login with prompt=none
    r -->> u: redirect to op
    u ->> o: access Authorization Endpoint
    o ->> o: check session
    note over o: session is valid. user is authenticated
    o ->> o: constructing the callback URL
    o -->> u: redirect to rp callback endpoint
    note over u: the following omitted
```

I believe that `constructing the callback URL` should be handled within [op.Authorize](https://github.com/zitadel/oidc/blob/24869d281155064b0a4a6339ad641ba76bac6177/pkg/op/server.go#L67). If there are any implementation examples or guidance on this, I would like to refer to them.

**Reply by hifabienne** [reply]
Let me tag @muhlemmer and @livio-a I think they are able to help you better than me

**Reply by muhlemmer** [reply, marked as answer]
In our product [ZITADEL](https://github.com/zitadel/zitadel) we pin the useragent to the user. When there is exactly one SSO session active, we set the userinfo when creating the Auth Request:

https://github.com/zitadel/zitadel/blob/6780c5a07ca491690e0af6d8baeac9aa5d69cabe/internal/auth/repository/eventsourcing/eventstore/auth_request.go#L173-L176

https://github.com/zitadel/zitadel/blob/6780c5a07ca491690e0af6d8baeac9aa5d69cabe/internal/auth/repository/eventsourcing/eventstore/auth_request.go#L740-L757

At that point the user is redirected to the login UI URL. The login UI will retrieve the auth request and sees prompt == none. Then it will use the `/auth/callback` URI to send the user back the the OIDC library.

https://github.com/zitadel/zitadel/blob/6780c5a07ca491690e0af6d8baeac9aa5d69cabe/internal/auth/repository/eventsourcing/eventstore/auth_request.go#L1034-L1037

The AuthorizeCallback handler then must check all requirements are fulfilled:

https://github.com/zitadel/oidc/blob/9f7cbb0dbfc39ce85ecb2f2024d6f20d44e8fcef/pkg/op/auth_request.go#L418-L440

Note that the actual checks happen in the GetAuthRequestByID. In our implementation the above `nextSteps` function is called again, but this time the `checkLoggedIn` flag is `true`. This time we check there is a userID in the auth request and that the user is still active etcetera:

https://github.com/zitadel/zitadel/blob/6780c5a07ca491690e0af6d8baeac9aa5d69cabe/internal/auth/repository/eventsourcing/eventstore/auth_request.go#L1038-L1047

**Reply by nannany** [reply]
Thank you for providing the implementation example.

Based on the above example, I understand that the sequence would proceed as follows:

```mermaid
sequenceDiagram
    participant u as User Agent
    participant r as Relying Party
    participant o as OpenID Provider
    note over u: omitted
    u ->> o: access Authorization Endpoint with prompt=none
    o ->> o: check session
    note over o: session is valid. user is authenticated
    o -->> u: redirect to /auth/callback
    u ->> o: access /auth/callback
    o -->> u: redirect to relying party's redirect_uri
    u ->> r: access redirect_uri
    note over u: the following omitted
```

-----

I understand that the `/auth/callback` endpoint on the OpenID Provider is outside the scope of the OIDC specification and is a unique mechanism specific to `zitadel/oidc`.

From reviewing [the proposal](https://github.com/zitadel/oidc/issues/440), it appears that support for `/auth/callback` will be removed in version 4. Do you have any expectations or recommendations on how to handle this change when it occurs?

**Reply by muhlemmer** [reply]
Well, it should be possible for a login implementation to redirect the user directly to the application upon fulfilling the last step.
As mentioned in the issue, because of the storage interface requirement, we needed the redirect back to OIDC to mark the auth request as completed.

In the future we want to use the `Server` interface instead. In that case the handling of the auth request and related storage can be implemented freely by the consumer and it's easier to complete the flow.
